closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -318,16 +354,86 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
         (NULL == OSSL_RSA_private_encrypt) ||
         (NULL == OSSL_BN_bin2bn) ||
         (NULL == OSSL_BN_set_negative) ||
-        (NULL == OSSL_BN_free)) {
-        //fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
-        //fflush(stderr);
+        (NULL == OSSL_BN_free) ||
+        ((NULL == OSSL_CRYPTO_num_locks) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_numeric) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_OPENSSL_malloc) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_callback) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_set_locking_callback) && (0 == ossl_ver))) {
+        // fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
+        // fflush(stderr);
         unload_crypto_library(handle);
         return -1;
     } else {
+        if (0 == ossl_ver) {
+            if (0 != thread_setup()) {
+                unload_crypto_library(handle);
+                return -1;
+            }
+        }
         return 0;
     }
  }
 
+#ifdef WINDOWS
+static HANDLE *lock_cs;
+
+int thread_setup()
+{
+    int i;
+    lock_cs = (*OSSL_OPENSSL_malloc)((*OSSL_CRYPTO_num_locks)() * sizeof(HANDLE));
+    if (NULL == lock_cs) {
+        return -1;
+    }
+    for (i = 0; i < (*OSSL_CRYPTO_num_locks)(); i++) {
+        lock_cs[i] = CreateMutex(NULL, FALSE, NULL);
+    }
+    (*OSSL_CRYPTO_set_locking_callback)((void (*)(int, int, char *, int))
+                                win32_locking_callback);
+    return 0;
+}
+
+void win32_locking_callback(int mode, int type, const char *file, int line)
+{
+    if (0 != (mode & CRYPTO_LOCK)) {
+        WaitForSingleObject(lock_cs[type], INFINITE);
+    } else {
+        ReleaseMutex(lock_cs[type]);
+    }
+}
+
+#else
+static pthread_mutex_t *lock_cs;
+
+int thread_setup()
+{
+    int i;
+    lock_cs = (*OSSL_OPENSSL_malloc)((*OSSL_CRYPTO_num_locks)() * sizeof(pthread_mutex_t));
+    if (NULL == lock_cs) {
+        return -1;
+    }
+    for (i = 0; i < (*OSSL_CRYPTO_num_locks)(); i++) {
+        pthread_mutex_init(&(lock_cs[i]), NULL);
+    }
+    (*OSSL_CRYPTO_THREADID_set_callback)(pthreads_thread_id);
+    (*OSSL_CRYPTO_set_locking_callback)(pthreads_locking_callback);
+    return 0;
+}
+
+void pthreads_locking_callback(int mode, int type, const char *file, int line)
+{
+    if (0 != (mode & CRYPTO_LOCK)) {
+        pthread_mutex_lock(&(lock_cs[type]));
+    } else {
+        pthread_mutex_unlock(&(lock_cs[type]));
```

Comment:
Like for Windows, the values returned from `pthread_mutex_lock()` and `pthread_mutex_unlock()` should be checked.